PublicMapping/districtbuilder
# Pin metric improvements

## Overview

Brief description of what this PR does, and why it is needed.

### Checklist


### Demo

Pinning lots of fields, and the updates don't mess w/ each other:
![Animation](https://user-images.githubusercontent.com/4432106/130808140-26ba9856-0c7b-4dc0-985d-70001a9a25ce.gif)

Pinning fields on a read-only project:
![Animation2](https://user-images.githubusercontent.com/4432106/130808432-e49139f2-fc99-4b3c-98c9-f74b1acff72f.gif)


### Notes

This worked out to be easier than anticipated because a lot of the redux code was already written, but was not being used 😳 .

I originally only meant to fix #956 , but including #898 as well turned out to be relatively simple after I had refactored the reducer code.

## Testing Instructions

- Open a project and pin many fields in quick succession. On `develop` / staging your pins will get overwritten as updates complete, but on this PR that won't happen
  - Reloading the page show show that your pins have persisted
- Open another users project while logged in, or open a project while not logged in (e.g. from an incognito window). You should be able to pin fields - on `develop` it will fail w/ a server error.
  - reloading the project should _not_ persist your pins
  
Closes #956 
Closes #898
